lib/rpc/broker.toit

```diff
@@ -134,3 +155,34 @@ monitor RpcRequestQueue_:
       if identical last_ request: last_ = next
       first_ = next
       return request
+
+  cancel id/int -> int:
+    // For testing purposes, we keep track of the number of requests that
+    // were actually canceled through this operation.
+    result/int := 0
+    // First we get rid of any unprocessed request with the given id. This
+    // is a simple linked list traversal with the usual bookkeeping challenges
+    // that come from removing from a linked list with insertion at the end.
+    previous := null
+    current := first_
+    while current:
+      next := current.next
+      if current.id == id:
+        if previous:
+          previous.next = next
+        else:
+          first_ = next
+        if not next:
+          last_ = previous
+        unprocessed_--
+        result++
+      previous = current
+      current = next
+    // Then we cancel any requests that are in progress by canceling the
+    // associated processing task.
+    processing_requests_.size.repeat:
+      request/RpcRequest_? := processing_requests_[it]
+      if request and request.id == id:
+        processing_tasks_[it].cancel
```

Comment:
Is there any change that we leave requests unprocessed? Normally tasks stay alive until there is no more work to do, but if all worker tasks are cancelled then we may have unprocessed request left in the pool that won't be processed until a new worker task is created.

Reply:
No, I don't think so, but that's because of a bug: Canceled processing tasks stick around, but don't do much processing after that. I think I've worked around that issue.

Reply:
That is: Canceled processing tasks are now removed from the processing pool and new tasks are spawned on demand.